bleach/html5lib_shim.py

```diff
@@ -37,8 +38,156 @@
+    where Bleach's cleaner is going to escape a disallowed tag so we can
+    escape the original string.
+
+    """
+    def __init__(self, inner_stream):
+        self._inner_stream = inner_stream
+        self._buffer = []
+
+    @property
+    def errors(self):
+        return self._inner_stream.errors
+
+    def reset(self):
+        return self._inner_stream.reset()
+
+    def position(self):
+        return self._inner_stream.position()
+
+    def char(self):
+        c = self._inner_stream.char()
+        # char() can return None if EOF, so ignore that
+        if c:
+            self._buffer.append(c)
+        return c
+
+    def charsUntil(self, characters, opposite=False):
+        chars = self._inner_stream.charsUntil(characters, opposite=opposite)
+        self._buffer.extend(list(chars))
+        return chars
+
+    def unget(self, char):
+        if self._buffer:
+            self._buffer.pop(-1)
+        return self._inner_stream.unget(char)
+
+    def stream_history(self):
+        return self._buffer
+
+
+def get_recent_tag_string(stream_history, token):
+    """Find the original text for the tag
+
+    This goes back through the stream we've tokenized for the most recent
+    complete HTML tag-like thing as it existed in the stream. It assumes that
+    the current character in the stream ias a >.
+
+    :arg list stream_history: list of characters to look through
+    :arg dict token: the tag token we're looking for in the stream
+
+    :returns: original tag from < to >
+
+    """
+    name_reversed = list(reversed(token['name']))
+    if token['type'] == tokenTypes['EndTag']:
+        name_reversed.append('/')
+    name_reversed_len = len(name_reversed)
+
+    quotey_things = '"\''
+    pile = []
+    in_quotes = []
+    for c in reversed(stream_history):
+        if in_quotes:
+            if c == in_quotes[-1]:
+                in_quotes.pop(-1)
+            elif c in quotey_things:
+                in_quotes.append(c)
+
+            pile.append(c)
+
+        elif c in quotey_things:
+            in_quotes.append(c)
+            pile.append(c)
+
+        elif c == '<':
+            if pile[-name_reversed_len:] == name_reversed:
+                pile.append(c)
+                break
+            else:
+                pile.append(c)
+        else:
+            pile.append(c)
+
+    pile.reverse()
+    ret = six.text_type('').join(pile)
+    return ret
+
+
 class BleachHTMLTokenizer(HTMLTokenizer):
     """Tokenizer that doesn't consume character entities"""
+    def __init__(self, stream, parser=None, **kwargs):
+        # Stomp on the HTMLTokenizer __init__ in order to wrap the stream.
+        self.stream = InputStreamWithMemory(HTMLInputStream(stream, **kwargs))
+
+        # Do all the things the HTMLTokenizer does in __init__
+        self.parser = parser
+        self.escapeFlag = False
+        self.lastFourChars = []
+        self.state = self.dataState
+        self.escape = False
+        self.currentToken = None
```

Comment:
This `__init__` is the same as `HTMLTokenizer` except it wraps the input stream with one that remembers the history of the stream.

Reply:
Oh, that's a really good idea. I think I didn't do that because there was an intermediary period where this was doing more. I forget.

I'll fix this now.